### PR TITLE
chain: DataSource handler

### LINKED ISSUE #496: Implement handler for data sources

### PR DESCRIPTION
**Pull request by evilpeach**
Fixes #496

### CODE REVIEW

File: chain/x/zoracle/handler_test.go [hunk 1]
```diff
@@ -480,3 +484,57 @@ func setupTestValidator(ctx sdk.Context, keeper Keeper, pk string) sdk.ValAddres
 // 	pendingRequests = keeper.GetPendingRequests(ctx)
 // 	require.Equal(t, []uint64{}, pendingRequests)
 // }
+
+func mockDataSource(ctx sdk.Context, keeper Keeper) {
```
Comment (taobun) [inline chain/x/zoracle/handler_test.go]: Return sdk.Result

File: chain/x/zoracle/handler_test.go [hunk 2]
```diff
@@ -480,3 +484,57 @@ func setupTestValidator(ctx sdk.Context, keeper Keeper, pk string) sdk.ValAddres
...
+	owner := sdk.AccAddress([]byte("owner"))
```
Comment (taobun) [inline chain/x/zoracle/handler_test.go:488]: Reuse code?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/handler.go [hunk 3]
```diff
@@ -148,6 +153,39 @@ func NewHandler(keeper Keeper) sdk.Handler {
 // 	return sdk.Result{Events: ctx.EventManager().Events()}
 // }
 
+func handleMsgCreateDataSource(ctx sdk.Context, keeper Keeper, msg MsgCreateDataSource) sdk.Result {
+	newDataSourceID := keeper.GetNextDataSourceID(ctx)
+
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: Add TODO: check executable size.

File: chain/x/zoracle/handler.go [hunk 4]
```diff
@@ -148,6 +153,39 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	newDataSource := types.NewDataSource(
+		msg.Owner,
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: Check that fee contains only valid denom

File: chain/x/zoracle/handler.go [hunk 5]
```diff
@@ -148,6 +153,39 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	return sdk.Result{}
+}
+
+func handleMsgEditDataSource(ctx sdk.Context, keeper Keeper, msg MsgEditDataSource) sdk.Result {
+	// Check DataSourceID is exist.
+	dataSource, err := keeper.GetDataSource(ctx, msg.DataSourceID)
+	if err != nil {
+		return err.Result()
+	}
+
```
Comment (sorawit) [inline chain/x/zoracle/handler.go:171]: Check that sender is the original owner.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/handler.go [hunk 6]
```diff
@@ -148,6 +153,39 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	newDataSource := types.NewDataSource(
+		msg.Owner,
+		msg.Name,
+		msg.Fee,
+		msg.Executable,
+	)
+
+	// Save DataSource to state.
+	keeper.SetDataSource(ctx, newDataSourceID, newDataSource)
+	return sdk.Result{}
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: Let's also return events here

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/handler.go (hunk 4 above)
Comment (evilpeach) [reply chain/x/zoracle/handler.go]: another module doesn't check it

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 7]
```diff
@@ -6,9 +6,12 @@ import (
 )
 
 // SetDataSource saves the given data source with the given ID to the storage.
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/data_source.go:8]:
```
func (k Keeper) AddDataSource(owner sdk.AccAddress, name string, executable []byte) sdk.Error {
   ...
}
```

File: chain/x/zoracle/handler.go [hunk 8]
```diff
@@ -148,6 +153,48 @@ func NewHandler(keeper Keeper) sdk.Handler {
 // 	return sdk.Result{Events: ctx.EventManager().Events()}
 // }
 
+// handleMsgCreateDataSource is a function to handle MsgCreateDataSource.
+func handleMsgCreateDataSource(ctx sdk.Context, keeper Keeper, msg MsgCreateDataSource) sdk.Result {
+	newDataSourceID := keeper.GetNextDataSourceID(ctx)
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]:
```
err := keeper.AddDataSource(msg.owner, ...)
if err != nil { 
   return err.Result()
}
return ...
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/handler.go [hunk 9]
```diff
@@ -148,6 +153,34 @@ func NewHandler(keeper Keeper) sdk.Handler {
...
+	if err != nil {
+		return err.Result()
+	}
+
+	if !dataSource.Owner.Equals(msg.Sender) {
+		// TODO: change it later.
+		return types.ErrInvalidOwner(types.DefaultCodespace).Result()
+	}
+
+	err = keeper.EditDataSource(ctx, msg.Owner, msg.DataSourceID, msg.Name, msg.Fee, msg.Executable)
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: let's put ID as the first argument.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]
Thanks I think this looks great!

File: chain/x/zoracle/internal/keeper/data_source.go [hunk 10]
```diff
@@ -5,10 +5,26 @@ import (
...
+
+// EditDataSource edits the given data source by given data source id to the storage.
+func (k Keeper) EditDataSource(ctx sdk.Context, dataSourceID int64, owner sdk.AccAddress, name string, fee sdk.Coins, executable []byte) sdk.Error {
+	// TODO: check executable size.
+	updatedDataSource := types.NewDataSource(owner, name, fee, executable)
+
 	store := ctx.KVStore(k.storeKey)
-	store.Set(types.DataSourceStoreKey(id), k.cdc.MustMarshalBinaryBare(dataSource))
+	store.Set(types.DataSourceStoreKey(dataSourceID), k.cdc.MustMarshalBinaryBare(updatedDataSource))
+	return nil
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/data_source.go:37]: Both function have duplicate code to store dataSource. I think we have `SetDataSource` function to manage key and save to store and 2 functions can call it

Comment (evilpeach) [reply chain/x/zoracle/internal/keeper/data_source.go:37]: private?

Comment (taobun) [reply chain/x/zoracle/internal/keeper/data_source.go:37]: public is ok i think

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**